walrus-attributes-indexer/src/handlers/blog_post.rs

```diff
@@ -0,0 +1,340 @@
+                }
+                ProcessedWalrusMetadata::Delete(dynamic_field_id) => {
+                    batch.insert(dynamic_field_id, value);
+                }
+            }
+        }
+    }
+
+    async fn commit<'a>(batch: &Self::Batch, conn: &mut postgres::Connection<'a>) -> Result<usize> {
+        // Partition the batch into items to delete and items to upsert
+        let (to_delete, to_upsert): (Vec<_>, Vec<_>) = batch
+            .values()
+            .partition(|item| matches!(item, ProcessedWalrusMetadata::Delete(_)));
+
+        let to_upsert: Vec<StoredBlogPost> = to_upsert
+            .into_iter()
+            .map(|item| item.into_stored())
+            .collect::<Result<Vec<_>>>()?;
+
+        let to_delete: Vec<ObjectID> = to_delete
+            .into_iter()
+            .map(|item| Ok(item.dynamic_field_id()))
+            .collect::<Result<Vec<_>>>()?;
+
+        let mut total_affected = 0;
+
+        if !to_delete.is_empty() {
+            let deleted_count = diesel::delete(blog_post::table)
+                .filter(blog_post::dynamic_field_id.eq_any(to_delete.iter().map(|id| id.to_vec())))
+                .execute(conn)
+                .await?;
+
+            total_affected += deleted_count;
+        }
+
+        if !to_upsert.is_empty() {
+            let upserted_count = diesel::insert_into(blog_post::table)
+                .values(&to_upsert)
+                .on_conflict(blog_post::dynamic_field_id)
+                .do_update()
+                .set((
+                    blog_post::df_version.eq(excluded(blog_post::df_version)),
+                    blog_post::owner_id.eq(excluded(blog_post::owner_id)),
+                    blog_post::title.eq(excluded(blog_post::title)),
+                    blog_post::blob_id.eq(excluded(blog_post::blob_id)),
+                    blog_post::view_count.eq(excluded(blog_post::view_count)),
+                ))
+                .filter(blog_post::df_version.lt(excluded(blog_post::df_version)))
+                .execute(conn)
+                .await?;
+
+            total_affected += upserted_count;
+        }
+
+        Ok(total_affected)
+    }
+}
+
+impl FieldCount for ProcessedWalrusMetadata {
+    const FIELD_COUNT: usize = StoredBlogPost::FIELD_COUNT;
+}
+
+impl BlogPostPipeline {
+    pub fn new(type_string: &str) -> Result<Self> {
+        let metadata_type = parse_sui_struct_tag(type_string)?;
+        Ok(BlogPostPipeline { metadata_type })
+    }
+}
+
+impl ProcessedWalrusMetadata {
+    fn dynamic_field_id(&self) -> ObjectID {
+        match self {
+            ProcessedWalrusMetadata::Upsert {
+                dynamic_field_id, ..
+            } => *dynamic_field_id,
+            ProcessedWalrusMetadata::Delete(dynamic_field_id) => *dynamic_field_id,
+        }
+    }
+
+    /// Attempt to convert into a `StoredBlogPost` only if the variant is `Upsert`.
+    fn into_stored(&self) -> Result<StoredBlogPost> {
```

Comment:
nit: `into` implies consumption,

```suggestion
    fn to_stored(&self) -> Result<StoredBlogPost> {
```